google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/ManagedCloudSdkService.java

```diff
@@ -16,35 +16,234 @@
 
 package com.google.cloud.tools.intellij.appengine.sdk;
 
+import com.google.cloud.tools.intellij.util.ThreadUtil;
+import com.google.cloud.tools.managedcloudsdk.ManagedCloudSdk;
+import com.google.cloud.tools.managedcloudsdk.ManagedSdkVerificationException;
+import com.google.cloud.tools.managedcloudsdk.ManagedSdkVersionMismatchException;
+import com.google.cloud.tools.managedcloudsdk.MessageListener;
+import com.google.cloud.tools.managedcloudsdk.UnsupportedOsException;
+import com.google.cloud.tools.managedcloudsdk.command.CommandExecutionException;
+import com.google.cloud.tools.managedcloudsdk.command.CommandExitException;
+import com.google.cloud.tools.managedcloudsdk.components.SdkComponent;
+import com.google.cloud.tools.managedcloudsdk.install.SdkInstallerException;
+import com.google.cloud.tools.managedcloudsdk.install.UnknownArchiveTypeException;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Lists;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.diagnostic.Logger;
+import java.io.IOException;
 import java.nio.file.Path;
+import java.util.Collection;
 import org.jetbrains.annotations.Nullable;
 
 /**
  * {@link CloudSdkService} providing SDK that is managed and automatically installable, see {@link
- * com.google.cloud.tools.managedcloudsdk.ManagedCloudSdk}
+ * com.google.cloud.tools.managedcloudsdk.ManagedCloudSdk}.
+ *
+ * <p>Some methods ({@link #install()} do their work on background thread but listeners methods are
+ * always called on EDT thread. This class is not thread-safe, must be used on UI thread.
  */
-// TODO(ivanporty) implementation coming in the next PR
 public class ManagedCloudSdkService implements CloudSdkService {
+  private Logger logger = Logger.getInstance(ManagedCloudSdkService.class);
+
+  private ManagedCloudSdk managedCloudSdk;
+
+  private SdkStatus sdkStatus = SdkStatus.NOT_AVAILABLE;
```

Comment:
its a little odd to see the status being set like this as a member variable. What happens if the SDK was already installed; what updates the status to READY? I wonder if this can be refactored a bit for clarity.

Reply:
Status is updated to `READY` when `activate()` is called, so before that we're not checking SDK and assume we're not available to avoid extra work and possible errors. Maybe move this assignment to constructor?

Reply:
so just at a higher level:

when is activate called: Lets say the user is using a managed SDK. Its installed, and activated, and therefore the status is updated to READY. What happens the next time the user boots up the IDE from scratch?

Reply:
So when IDE is booted, plugin init code is called, and it checks which SDK service is active, then creates it and calls `activate()`. SDK service checks its status.
Initially I was thinking about doing this in a constructor but it's a lot of work which is not what constructors are usually for.

Reply:
can you point me to that initialization code?

Reply:
or is that coming next?

Reply:
Correct, in next PRs, something like:
```java
CloudToolsPluginInitializationComponent {
...
SdkService sdkService = createSdkServiceFromSettings(...);
sdkService.activate();
CloudSdkService.setInstance(sdkService);
```

Reply:
got it. so just to confirm: active / install will have to be expanded in the next PRs to initialize the status (and skip install if already READY)?

Reply:
correct, init code will just do `activate()`, and Managed SDK service will skip install if current SDK appears to be valid (i.e. isInstalled() == true)

Reply:
sounds good. im ok with this how it is for now then. Will take another look in the next PRs when its expanded.